Implement unpack_at for JGF

milroy opened the pull request:
This PR implements the `unpack_at()` function in the JGF reader as discussed in issue #767.  The work in this PR is a necessary component of `grow` with the RV1/JGF reader and complements PR #773. It also adds a subcommand to the `resource-query` utility to enable testing of `unpack_at()`.  Once more of the infrastructure is in place, we can add `unpack_at()` tests to `flux-ion-resource` as well.

`Unpack_at()` requires a vertex attachment point and a fully-qualified JGF subgraph as inputs.  In this context, fully qualified means that the subgraph contains the root vertex and all intermediate vertices between the root and the new vertices to be attached.  In other words, every vertex in the subgraph has a path to the root.  The implementation is robust and doesn't assume any ordering of the vertices or edges.  It iterates through the JGF subgraph vertices and adds vertices not present in the resource graph, keeping track of the added vertex ids with an `std::unordered_set`.  When iterating through the subgraph edges, it only adds edges where either the source or target vertices (or both) have been added, ensuring that pre-existing edges aren't re-added to the resource graph.

Note: as indicated in the conversation below, this functionality is currently experimental as resource graph growth causes a resize of the boost `vecS` vertex container type.  Resizing the `vecS` results in lost job allocations and reservations as there is no copy constructor for planner.  Furthermore, to initialize the pruning filters of the new JGF subgraph, the implementation developed here reinitializes the entire resource graph.  We intend to add a performance optimizing PR to improve the pruning filter initialization in the future. 

This PR addresses and solves issues #782 and #783.

milroy commented:
Note that I will be adding additional tests to the testsuite as this PR evolves.

milroy commented:
@dongahn I've been creating more tests for CI and have a question about handling containment subsystem `in` edges.

I've been assuming that the JGF subgraph to be attached will have both `contains` and `in` edges.  However, if I pass a JGF with bidirectional edges to `update attach`, the traverser update here: https://github.com/flux-framework/flux-sched/blob/97b91cccc4b29a41cad41c3ab54bb9c24e811b3f/resource/utilities/command.cpp#L258
results in a segfault.  Is this expected behavior?  If so, should I handle it in `resource-query` by passing two JGFs to `update attach`: one with `contains` and `in` edges for `unpack_at ()`, and one with only `contains` edges for the `traverser->run ()` step?

dongahn commented:
@milroy: As we discussed, I am going to have a review pass with a goal to land this right after PR #773. Could you remove [WIP] label when you are ready?

dongahn commented:
> If so, should I handle it in resource-query by passing two JGFs to update attach: one with contains and in edges for unpack_at (), and one with only contains edges for the traverser->run () step?

No you shouldn't have to. There seems to be a logic in `run ()` that doesn't like a bidirectional edge. We should look into it.

dongahn commented:
If you pass a bidirectional JGF into the existing interface (plain `update`), does `run()` still crash?

milroy commented:
> If you pass a bidirectional JGF into the existing interface (plain `update`), does `run()` still crash?

Yes, it does crash:
```bash
$ resource/utilities/resource-query -L t/data/resource/jgfs/elastic/l2-full.json -F jgf -f jgf -S CA -P high
INFO: Loading a matcher: CA
resource-query> update allocate t/data/resource/jgfs/elastic/node-add-test.json 1 0 3600
Segmentation fault
```

dongahn commented:
Could you create an issue for this? Since this is a distinct problem, it would be ideal to address it as a separate PR then yours?

milroy commented:
@dongahn no problem!  I'll create an issue and get to work on a PR.

milroy commented:
>  I'll create an issue and get to work on a PR.

However, fixing the issue enables CI testing for `unpack_at ()`, so I can see an argument for fixing it in this PR.

milroy commented:
@dongahn I opened issue #782 to discuss the segfault, but the fix appears to be a one liner.  If that looks good to you I can add a commit to this PR.

dongahn commented:
Yes, that sounds great!

milroy commented:
I ran into another problem with the traverser update step and created issue #783 for discussion.  That issue needs to be solved before this PR can be reviewed.

milroy commented:
I've dropped the WIP tag and have added two CI tests.  As soon as I add the remaining tests this PR will be ready for review.

milroy commented:
This should be ready for your review, @dongahn and @SteVwonder.

milroy commented:
I forgot to add `find sched-now=allocated` to CI test 7, so I committed two fixups and rebased.  The PR is actually ready for review now.

milroy commented:
@dongahn and @SteVwonder: please hold off on your reviews.  While preparing the demo for our collaborators, I ran into a problem.

The `unpack_at ()` implementation in this PR currently clears all allocations from the resource graph while attaching the new subgraph.  I'm currently investigating this issue and will update the PR as soon as I have more information.

milroy commented:
I determined that when new vertices are added to the resource graph, all allocations are cleared.  

@dongahn and I discussed this during the weekly team meeting.  He suspected this problem is caused by the lack of a copy constructor for boost `vecS` vertices.  I've confirmed that preallocating a large number of vertices before populating the resource graph prevents this problem, which is consistent with the lack of copy constructor for vertices.

Let's pause this PR for now as we will need further, thorough investigation to determine how best to fix the problem.  I may delete this PR until we have a good solution.

dongahn commented:
Thanks @milroy. Just to be clear, please see https://github.com/flux-framework/flux-sched/blob/master/resource/schema/sched_data.cpp#L38

I mainly chose to do this because two reasons: 1) because there is no grow support needed, there will be no dynamic resizing and the schedule state won't be copied once the graph is initialized; 2) there is no "copy constructor" in planner (though it is C but it is a class style): https://github.com/flux-framework/flux-sched/blob/master/resource/planner/planner.h.

As we will want to be able to grow and shrink beyond the initialization in a future, it seems we will need to design this and break down the tasks a bit. 

I am also wondering if we can still land most of this PR by limiting the use case of this... (Let's chat)

dongahn commented:
> @dongahn and @SteVwonder: please hold off on your reviews. While preparing the demo for our collaborators, I ran into a problem.

@milroy: I will look at your PR tomorrow if you are okay. I want to see if some portion of this code can go in sooner rather than later because I suspect we will need it soon.

dongahn commented on t/data/resource/commands/elastic/cmds02.in:
```diff
@@ -0,0 +1,5 @@
+find status=up
+update attach @TEST_SRCDIR@/data/resource/jgfs/elastic/node-add-test-random.json 1 0 3600
+find status=up
+quit
```
Interestingly, my GitHub renderer added an error icon with "No new line at end of file".
<img width="739" alt="Screen Shot 2020-12-17 at 11 33 03 PM" src="https://user-images.githubusercontent.com/862123/102587144-3e6a8980-40c0-11eb-96c8-c27b4e1db7c3.png">

dongahn commented on resource/utilities/command.cpp:
```diff
@@ -48,7 +48,7 @@ command_t commands[] = {
 "resource-query> match allocate jobspec"},
     { "update", "u", cmd_update, "Update resources with a JGF subgraph (subcmd: "
-"allocate | reserv): "
+"allocate | reserv | attach): "
```
It is not a part of you commit but I just see a typo in `reserv`. Perhaps you can correct it while you are on this line.

milroy replied:
Will do!

dongahn commented on resource/readers/resource_reader_jgf.cpp:
```diff
@@ -293,8 +294,14 @@ int resource_reader_jgf_t::add_graph_metadata (vtx_t v,
             ptr = m.roots.emplace (kv.first, v);
-            if (!ptr.second)
+            if (!ptr.second) {
+                errno = EPROTO;
```
This is my fault. But errno should change to `EINVAL`. I just have a global whole sale change in my PR #787.

milroy replied:
No problem!

dongahn commented on resource/readers/resource_reader_jgf.cpp:
```diff
@@ -551,6 +558,63 @@ int resource_reader_jgf_t::unpack_vertices (resource_graph_t &g,
     return rc;
 }
 
+int resource_reader_jgf_t::unpack_vertices (resource_graph_t &g,
```
First a cursory question before taking a deeper dive: I see you overloaded `unpack_vertices` and `unpack_edges`. How difficult would it be to unified these functions so that we can use one version for both `unpack` and `unpack_at`?

milroy replied:
During our conversation yesterday we determined it wouldn't be difficult, and my tested solution indeed is straightforward.

dongahn reviewed (commented):
I had a bit of time so I had a cursory review. BTW, I really like how you structured your commits so that each commit only has one topic!

I will look more tomorrow but I wonder if this PR can go in sooner with having a label like "experimental feature" somewhere. Since it will just not be used in production use, it seems to make sense. Let's chat tomorrow.

milroy commented:
> I will look more tomorrow but I wonder if this PR can go in sooner with having a label like "experimental feature" somewhere.

Should I update the PR title with [Experimental]?

@dongahn I believe I have addressed your comments, and have updated the PR to reflect them.  You can take another look whenever you get a chance.  

@SteVwonder go ahead with your review when you're ready.

milroy commented:
> First a cursory question before taking a deeper dive: I see you overloaded `unpack_vertices` and `unpack_edges`. How difficult would it be to unified these functions so that we can use one version for both `unpack` and `unpack_at`?

After making this change `unpack` and `unpack_at` are exactly the same (except for the vertex argument passed to `unpack_at`, which isn't used).  Should we just get rid of the `unpack_at` interface and just always use `unpack`?

milroy commented:
I've rebased to pick up the latest changes to master and included the changes we discussed at coffee time.  This is ready for another review @dongahn and @SteVwonder.

dongahn commented on resource/readers/resource_reader_jgf.cpp:
```diff
@@ -907,24 +950,18 @@ int resource_reader_jgf_t::update_tgt_edge (resource_graph_t &g,
+             e = *ei;
+             found = true;
+             break;
+        }
```
Should this be a new commit (e.g., removing an additional indentation)

dongahn commented on resource/utilities/command.cpp:
```diff
@@ -48,8 +48,9 @@ command_t commands[] = {
     { "update", "u", cmd_update, "Update resources with a JGF subgraph (subcmd: "
-"allocate | reserv): "
-"resource-query> update allocate jgf_file jobid starttime duration" },
+"allocate | reserve | attach): "
```
Like we chatted over phone, I believe adding "attach" as a first-class command (i.e., not a subcommand of update) would make the semantics of the `update` command a bit cleaner.

milroy replied:
Sure thing!

dongahn commented on resource/utilities/command.cpp:
```diff
@@ -233,6 +235,22 @@ static int update_run (std::shared_ptr<resource_context_t> &ctx,
     }
 
     gettimeofday (&st, NULL);
+    if (attach) {
```
And with the first class `attach` command, we won't have to introduce this control flow complexity.

dongahn commented on resource/utilities/command.cpp:
```diff
@@ -260,10 +281,11 @@ static int update (std::shared_ptr<resource_context_t> &ctx,
     std::stringstream buffer{};
 
-    if (!(subcmd == "allocate" || subcmd == "reserve")) {
+    if (!(subcmd == "allocate" || subcmd == "reserve" || subcmd == "attach")) {
```
Same

dongahn commented on resource/utilities/command.cpp:
```diff
@@ -274,8 +296,9 @@ static int update (std::shared_ptr<resource_context_t> &ctx,
-        || ctx->reservations.find (jobid) != ctx->reservations.end ()) {
+    if ( (ctx->allocations.find (jobid) != ctx->allocations.end ()
+        || ctx->reservations.find (jobid) != ctx->reservations.end ())
+        && (subcmd != "attach")) {
```
same

dongahn commented on resource/utilities/command.cpp:
```diff
@@ -290,7 +313,11 @@ static int update (std::shared_ptr<resource_context_t> &ctx,
-    return update_run (ctx, args[2], buffer.str (), jobid, at, d);
+    // Attach is an experimental feature.
+    if (subcmd == "attach")
+        attach = true;
+
+    return update_run (ctx, args[2], buffer.str (), jobid, at, d, attach);
```
Same

dongahn commented on t/data/resource/jgfs/elastic/node-add-test-random.json:
```diff
@@ -0,0 +1 @@
+{"graph": {"nodes": [{"id": "359", "metadata": {"type": "core", "basename": "core", "name": "core30", "id": 30, "uniq_id": 359, "rank": -1, "exclusive": true, "unit": "", "size": 1, "paths": {"containment": "/medium0/rack0/node3/socket1/core30"}}}, {"id": "347", "metadata": {"type": "core", "basename": "core", "name": "core18", "id": 18, "uniq_id": 347, "rank": -1, "exclusive": true, "unit": "",  ...
```
My GitHub viewer shows an error as this doesn't have a new line.

milroy replied:
Ok, I'll double check the new lines in the JSON files.

dongahn commented on t/data/resource/jgfs/elastic/node-add-test.json:
```diff
@@ -0,0 +1 @@
+{"graph": {"nodes": [{"id": "329", "metadata": {"type": "core", "basename": "core", "name": "core0", "id": 0, "uniq_id": 329, "rank": -1, "exclusive": true, "unit": "", "size": 1, "paths": {"containment": "/medium0/rack0/node3/socket0/core0"}}}, {"id": "330", "metadata": {"type": "core", "basename": "core", "name": "core1", "id": 1, "uniq_id": 330, "rank": -1, "exclusive": true, "unit": "", "size ...
```
My GitHub viewer shows an error as this doesn't have a new line.

dongahn commented on t/data/resource/jgfs/elastic/node-test.json:
```diff
@@ -0,0 +1 @@
+{"graph": {"nodes": [{"id": "221", "metadata": {"type": "core", "basename": "core", "name": "core0", "id": 0, "uniq_id": 221, "rank": -1, "exclusive": true, "unit": "", "size": 1, "paths": {"containment": "/medium0/rack0/node0/socket0/core0"}}}, {"id": "222", "metadata": {"type": "core", "basename": "core", "name": "core1", "id": 1, "uniq_id": 222, "rank": -1, "exclusive": true, "unit": "", "size ...
```
My GitHub viewer shows an error as this doesn't have a new line.

dongahn commented on t/data/resource/commands/elastic/cmds01.in:
```diff
@@ -0,0 +1,5 @@
+match allocate @TEST_SRCDIR@/data/resource/jobspecs/elastic/test-ma-node.yaml
+find status=up
+update attach @TEST_SRCDIR@/data/resource/jgfs/elastic/node-add-test.json 1 0 3600
```
Once `attach` is hoisted as a first class command, this will need an adjustment

dongahn commented on t/data/resource/commands/elastic/cmds02.in:
```diff
@@ -0,0 +1,5 @@
+match allocate @TEST_SRCDIR@/data/resource/jobspecs/elastic/test-ma-node.yaml
+find status=up
+update attach @TEST_SRCDIR@/data/resource/jgfs/elastic/node-add-test-random.json 1 0 3600
```
same

dongahn commented on t/data/resource/commands/elastic/cmds03.in:
```diff
@@ -0,0 +1,4 @@
+find status=up
+update attach @TEST_SRCDIR@/data/resource/jgfs/elastic/node-add-test.json 1 0 3600
```
same

dongahn commented on t/data/resource/commands/elastic/cmds04.in:
```diff
@@ -0,0 +1,7 @@
+match allocate @TEST_SRCDIR@/data/resource/jobspecs/elastic/test-ma-node.yaml
+find status=up
+find sched-now=allocated
+update attach @TEST_SRCDIR@/data/resource/jgfs/elastic/node-add-test.json 1 0 3600
```
same

dongahn commented on t/data/resource/commands/elastic/cmds05.in:
```diff
@@ -0,0 +1,4 @@
+find status=up
+update attach @TEST_SRCDIR@/data/resource/jgfs/power.json 1 0 3600
```
same

dongahn commented on t/data/resource/commands/elastic/cmds06.in:
```diff
@@ -0,0 +1,6 @@
+find status=up
+find sched-now=allocated
+update attach @TEST_SRCDIR@/data/resource/jgfs/elastic/node-test.json 1 0 3600
```
same

dongahn commented:
@milroy: also, I think there is huge differences between being about to attach a subgraph (grow) before the traverser initialized the graph or after the initialization: https://github.com/flux-framework/flux-sched/blob/master/resource/traversers/dfu.hpp#L79. How are you handling the reinitialization (e.g., pruning filter installation etc) as part of the attach operation?

milroy commented:
> @milroy: also, I think there is huge differences between being about to attach a subgraph (grow) before the traverser initialized the graph or after the initialization: https://github.com/flux-framework/flux-sched/blob/master/resource/traversers/dfu.hpp#L79. How are you handling the reinitialization (e.g., pruning filter installation etc) as part of the attach operation?

Thanks for the feedback!  This is a good point, and I must admit I didn't consider the reinitialization.  However, I think I have a solution that's fairly straightforward and doesn't require reinitializing the whole resource graph.  It will require creating a temporary boost subgraph (as I tried to avoid before).  Then I can overload `initialize` which will take the original resource graph as an argument.  `Initialize` will call an overloaded `dfu_impl_t::prime_pruning_filter` which will traverse the *temporary* subgraph, but update or initialize `subplans` depending on whether the vertex is new or already existed in the *original* resource graph.  This will also require an addition to `planner_multi` that sets the new counts (a setter similar to `planner_multi_resource_total_by_type`).

Let's talk about this during a coffee time.

milroy commented on resource/readers/resource_reader_jgf.cpp:
```diff
@@ -737,19 +744,50 @@ int resource_reader_jgf_t::unpack_vertices (resource_graph_t &g,
-        if (add_vtx (g, m, vmap, fetcher) != 0)
-            goto done;
+         
+        bool found = false;
```
Since `&&` is guaranteed to short circuit, this should be safe.

milroy commented:
After my discussions with @dongahn I've decided to implement the simplest (yet inefficient) way to initialize the added resource subgraph.  The updates to this PR add the capability of reinitializing the entire resource graph, which includes the new subgraph.  Since the process recreates subtree plans, allocated and reserved jobs may be affected.  I've added a check in `resource-query` to exit if there are outstanding allocations or reservations.

This is ready for another review.

milroy commented:
I broke out a line-wrap change into a new commit and made some minor edits to commit messages.

dongahn commented on resource/readers/resource_reader_jgf.cpp:
```diff
@@ -737,18 +763,32 @@ int resource_reader_jgf_t::unpack_vertices (resource_graph_t &g,
-            goto done;
+
+        vtx_t v = boost::graph_traits<resource_graph_t>::null_vertex ();
+        if ( (v = vtx_in_graph (m, fetcher.paths)) == null_vtx) {
```
Perhaps you can add a simple comment like:
```c++
// If the graph doesn't have the vertex, add it
```

dongahn commented on resource/readers/resource_reader_jgf.cpp:
```diff
@@ -737,18 +763,32 @@ int resource_reader_jgf_t::unpack_vertices (resource_graph_t &g,
+        if ( (v = vtx_in_graph (m, fetcher.paths)) == null_vtx) {
+            if (add_vtx (g, m, vmap, fetcher) != 0)
+                goto done;
+            added_vtcs.insert (std::string (fetcher.vertex_id));
```
```suggestion
            auto res = added_vtcs.insert (std::string (fetcher.vertex_id));
            if (!res.second) {
                errno = EEXIST;
                goto done;
            }
```

milroy replied:
I'll be sure to implement this change.

dongahn commented on resource/readers/resource_reader_jgf.cpp:
```diff
@@ -737,18 +763,32 @@ int resource_reader_jgf_t::unpack_vertices (resource_graph_t &g,
+            if (add_vtx (g, m, vmap, fetcher) != 0)
+                goto done;
+            added_vtcs.insert (std::string (fetcher.vertex_id));
+        }
+        else {
```
```suggestion
        } else {
```

dongahn commented on resource/readers/resource_reader_jgf.cpp:
```diff
@@ -513,6 +535,30 @@ int resource_reader_jgf_t::add_graph_metadata (vtx_t v,
 
+int resource_reader_jgf_t::update_vmap (std::map<std::string, 
+                                                 vmap_val_t> &vmap,
+                                        const vtx_t &v, 
```
```suggestion
                                       vtx_t v, 
```

`vtx_t` is a native integer data type and this doesn't need to be `const` ref.

dongahn commented on resource/readers/resource_reader_jgf.cpp:
```diff
@@ -850,19 +892,22 @@ int resource_reader_jgf_t::unpack_edges (resource_graph_t &g,
-            g[e].idata.member_of[std::string (key)]
-                = std::string (json_string_value (value));
+        if ( (added_vtcs.count (source) == 1) 
+              || (added_vtcs.count (target) == 1)) {
```
Maybe you can add an one-line comment about this condition? 
```c++
// We only unpack the edge when it connects at least one newly added vertex 
```

dongahn commented on resource/readers/resource_reader_jgf.hpp:
```diff
@@ -104,11 +105,16 @@ class resource_reader_jgf_t : public resource_reader_base_t {
                     std::map<std::string, bool> &is_roots);
     int add_graph_metadata (vtx_t v, resource_graph_t &g,
                             resource_graph_metadata_t &m);
+    int update_vmap (std::map<std::string, vmap_val_t> &vmap, const vtx_t &v, 
```
```suggestion
    int update_vmap (std::map<std::string, vmap_val_t> &vmap, vtx_t v,
```

dongahn reviewed (approved):
@milroy: this LGTM! I have added a few minor comments, which would be good to be addressed before this is landed. But since they are minor, I approve this PR. 

@SteVwonder: would you want to take a look at this PR as well?

milroy commented:
@dongahn, thanks for the helpful suggestions!  I'll implement them and push the fixes. @SteVwonder let me know if you want to look this PR over as well.

dongahn commented:
> @dongahn, thanks for the helpful suggestions! I'll implement them and push the fixes. @SteVwonder let me know if you want to look this PR over as well.

@milroy: very nicely done and those position us very well for ultimately elasticity! Thanks you. 

@SteVwonder: it would be great to land this to our next release (Jan release). I believe this only needs just one more cursory release.

SteVwonder commented on resource/readers/resource_reader_jgf.cpp:
```diff
@@ -984,6 +1032,7 @@ int resource_reader_jgf_t::unpack (resource_graph_t &g,
     json_t *nodes = NULL;
     json_t *edges = NULL;
     std::map<std::string, vmap_val_t> vmap;
+    std::unordered_set<std::string> added_vtcs;
```
Is this used by later commits?  If not, can you elaborate on the addition of this into the unpack_* function interfaces?

milroy replied:
The `unordered_set` is used primarily here: https://github.com/flux-framework/flux-sched/blob/b8cce32e4251ea8c3f2394157410667543b2a4ef/resource/readers/resource_reader_jgf.cpp#L763 

and here: https://github.com/flux-framework/flux-sched/blob/b8cce32e4251ea8c3f2394157410667543b2a4ef/resource/readers/resource_reader_jgf.cpp#L880

for _O(1)_ insertion and lookups of the vertices in the new JGF subgraph that are not present in the original resource graph.  By keeping track of the new vertices, we can add only the new edges as well (i.e. those that connect at least one new vertex).  If we didn't do this, we would end up re-adding edges that already exist in the resource graph, with likely unintended consequences.  The `added_vtcs` will likely be used by future optimizations for traversing and initializing only the newly added subgraph (rather than the entire, updated resource graph).

SteVwonder commented on t/t3028-resource-grow.t:
```diff
@@ -0,0 +1,81 @@
+test_expect_success "${test007_desc}" '
+    sed "s~@TEST_SRCDIR@~${SHARNESS_TEST_SRCDIR}~g" ${cmds007} > cmds007 &&
+    sed "s~@TEST_SRCDIR@~${SHARNESS_TEST_SRCDIR}~g" ${err007} > 007.R.err &&
+    ${query} -L ${jgf} -F jgf -f jgf -S CA -P low -t 007.R.out \
```
In the case of a nonexistent JGF, I think `resource-query` should exit with a non-zero exit code, right?  (Same for invalid argument and trying to grow when there are allocations).  If so, then A) I think `resource-query` isn't exiting with non-zero status and needs to be updated to do so B) this test should use `test_must_fail`.

dongahn replied:
@SteVwonder: I can take this. `resource-query` is designed to be an interactive CLI, it will be difficult to cause it to return on this condition. Certain the new command can return -1, which then cause the CLI to exit but that probably isn't what users want. At interactive prompt, when a user sees the error message from the command, he/she may want to correct the error (e.g., provide the correct path) and retry...

SteVwonder reviewed (commented):
Thanks @milroy for putting this together.  This is a huge leap in functionality.  I'm super excited to see it in action :) Just two question below:

milroy commented:
Thanks for the review and comments @SteVwonder!  Let me know if you need more details and if you'd like me to make changes before this gets merged.

SteVwonder reviewed (approved):
LGTM! Thanks @milroy!

milroy commented:
> LGTM! Thanks @milroy!

Great, thanks to you and @dongahn!  I'm going to rebase and set MWP.

dongahn commented:
Great job @milroy and @SteVwonder. Thanks!
